### PR TITLE
Fix reflexive relationships

### LINKED ISSUE #85: Reflexive relations
Reflexive relations 0 .. 1 between the same entity are not currently supported.

Example schema 

```ts
export default class Comment extends Resource {
  static get type() {
    return 'comment';
  }

  static schema = {
    attributes: {
      body: String,
      created_at: String,
      updated_at: String,
      parentCommentId: String
    },
    relationships: {
      user: {
        type: () => User,
        belongsTo: true,
        foreignKeyName: "userId"
      },
      parentComment: {
        type: () => Comment,
        belongsTo: true,
        foreignKeyName: "parentCommentId"
      },
      thread: {
        type: () => Thread,
        belongsTo: true,
        foreignKeyName: "threadId"
      }
    }
  };
}
```

It errors out when I do this query: `/comments?include=parentComment` with this error output:

`JSONAPI-TS:  { error: table name "comments" specified more than once `

### PR DESCRIPTION
**Pull request by spersico**
This PR should be merged after #88, and closes #85 . As I don't know the actual order in which this PR, #88, and #87 are going to be merged, I left it as it is, and will fix it when it needs to.

----------------------
**UPDATE APRIL 25:**
Added fix so that **ALL** relationships are shown, even when they are not set, or a include filter is present.

image

-----------------------
This PR only fixes reflexive relationships where the FK is on the base model (I didn't touch the hasMany SQL query generator).

------------

Basically, this PR does two things:
- It changes the _belongsTo_ SQL generator, so that one of the tables has an alias, to avoid a DB error.
- It adds a filter to the **serializer**, because it seems that the change in the SQL makes possible the existence of undefined objects in the **data.relationships** property.

This is an example of the endpoint working
image

### CODE REVIEW

File: src/application.ts [hunk 1]
```diff
@@ -155,7 +155,7 @@ export default class Application {
     const resourceClass = await this.resourceFor(data.type);
     const schemaRelationships = resourceClass.schema.relationships;
 
-    Object.keys(data.relationships).forEach(relationshipName => {
+    Object.keys(data.relationships).filter(relationshipName => data.relationships[relationshipName]).forEach(relationshipName => {
```
Comment (joelalejandro) [inline src/application.ts]: @spersico These filters could simply be `.filter(Boolean)`

File: src/processors/knex-processor.ts [hunk 2]
```diff
@@ -255,12 +255,12 @@ export default class KnexProcessor<
       return query
         .join(
           this.tableName,
-          `${foreignTableName}.id`,
+          `base${foreignTableName}.id`,
```
Comment (joelalejandro) [inline src/processors/knex-processor.ts]: @spersico What does `base` mean? I know it works as a prefix, but I'd like to know why do we call it that way.

Comment (spersico) [reply src/processors/knex-processor.ts]: because.... I felt like it :-P

File: src/application.ts (hunk 1 above)
Comment (spersico) [reply src/application.ts]: okay, I'll give it a go

File: src/processors/knex-processor.ts (hunk 2 above)
Comment (spersico) [reply src/processors/knex-processor.ts]: I'll give it a more descriptive name, like... **other** or something like that, I'll check the generated sql string and see

File: src/application.ts (hunk 1 above)
Comment (spersico) [reply src/application.ts]: @joelalejandro in https://github.com/ebryn/jsonapi-ts/pull/90/files#diff-8f42631f8d7d8cbaf47074d3ea5536d1R173 that worked, but here it breaks. Surely because one gets the keys (Object.keys), and the other the values (Object.values).
Even then, remember that this PR overlaps the #88, so when that one gets merged, part of this code will change.
I'll re-review it again then.

Comment (joelalejandro) [reply src/application.ts]: It's not that important, we can iterate on this later, @spersico.

**Comment by ebryn** [comment]
@spersico Can you fix these merge conflicts?

**Comment by spersico** [comment]
> @spersico Can you fix these merge conflicts?

yep, I was expecting this. I'll see to it later
